crates/wasi-provider/src/lib.rs

```diff
@@ -83,6 +91,178 @@ impl<S: Store + Send + Sync> WasiProvider<S> {
+                .await;
+            match start_result {
+                Ok(handle) => {
+                    container_handles.insert(container.name().to_owned(), handle);
+                }
+                Err(e) => {
+                    return (container_handles, Some(e));
+                }
+            }
+        }
+        info!(
+            "All containers started for pod {:?}. Updating status",
+            pod.name()
+        );
+
+        (container_handles, None)
+    }
+
+    async fn run_container_to_completion(
+        &self,
+        status_receiver: &mut tokio::sync::watch::Receiver<kubelet::container::Status>,
+        container: &Container,
+    ) -> anyhow::Result<()> {
+        let result = Self::wait_for_terminated_status(status_receiver).await;
+        debug!("Init container {} terminated", container.name());
+        if result.succeeded {
+            Ok(())
+        } else {
+            Err(anyhow::anyhow!(
+                "Init container {} failed with message {}",
+                container.name(),
+                result.message
+            ))
+        }
+    }
+
+    async fn run_init_containers(
+        &self,
+        pod: &Pod,
+        client: &kube::client::Client,
+        modules: &mut HashMap<String, Vec<u8>>,
+        volumes: &HashMap<String, Ref>,
+    ) -> (
+        HashMap<
+            String,
+            kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>,
+        >,
+        Option<anyhow::Error>,
+    ) {
+        info!("Running init containers for pod {:?}", pod.name());
+        let mut container_handles = HashMap::new();
+        for container in pod.init_containers() {
+            let start_result = self
+                .start_container(&container, pod, client, modules, volumes)
+                .await;
+            match start_result {
+                Ok(handle) => {
+                    let mut status_receiver = handle.status(); // TODO: ugh but borrow checker
+                    container_handles.insert(container.name().to_owned(), handle);
+                    let run_result = self
+                        .run_container_to_completion(&mut status_receiver, &container)
+                        .await;
+                    if let Err(run_error) = run_result {
+                        return (container_handles, Some(run_error));
+                    }
+                }
+                Err(e) => {
+                    return (container_handles, Some(e));
+                }
+            }
+        }
+        info!("Finished running init containers for pod {:?}", pod.name());
+        (container_handles, None)
+    }
+
+    async fn wait_for_terminated_status(
+        status_receiver: &mut tokio::sync::watch::Receiver<kubelet::container::Status>,
+    ) -> ContainerTerminationResult {
+        loop {
+            let status = status_receiver.next().await;
+            if let Some(Status::Terminated {
```

Comment:
I think `while let` should still work here no?

Reply:
I couldn't see a way to make it work - it looked like a failed match would drop out of the loop rather than go round again, and the compiler certainly thought it would result in ways out of the function other than via the `return`.  I'd be happy to take this suggestion but I'd need a more specific pointer on the correct phrasing of the `while let` - sorry.